sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -0,0 +1,274 @@
+      None
+    } else {
+      // There must be only one plan at the last level, which contains all items.
+      assert(plansLastLevel.size == 1 && plansLastLevel.head._1.size == items.length)
+      Some(plansLastLevel.head._2.plan)
+    }
+  }
+
+  /** Find all possible plans in one level, based on previous levels. */
+  private def searchForLevel(level: Int): Unit = {
+    val foundPlansCurLevel = foundPlans(level)
+    var k = 1
+    var continue = true
+    while (continue) {
+      val otherLevel = level - k
+      if (k > otherLevel) {
+        // We can break from here, because when building a join from A and B, both A J B and B J A
+        // are handled.
+        continue = false
+      } else {
+        val joinPlansLevelK = foundPlans(k).values.toSeq
+        for (i <- joinPlansLevelK.indices) {
+          val curJoinPlan = joinPlansLevelK(i)
+
+          val joinPlansOtherLevel = if (k == otherLevel) {
+            // Both sides of a join are at the same level, no need to repeat for previous ones.
+            joinPlansLevelK.drop(i)
+          } else {
+            foundPlans(otherLevel).values.toSeq
+          }
+
+          joinPlansOtherLevel.foreach { otherJoinPlan =>
+            // Should not join two overlapping item sets.
+            if (curJoinPlan.itemIds.intersect(otherJoinPlan.itemIds).isEmpty) {
+              val joinPlan = buildJoin(curJoinPlan, otherJoinPlan)
+              if (joinPlan.nonEmpty) {
+                // Check if it's the first plan for the item set, or it's a better plan than
+                // the existing one due to lower cost.
+                val existingPlan = foundPlansCurLevel.get(joinPlan.get.itemIds)
+                if (existingPlan.isEmpty || joinPlan.get.cost < existingPlan.get.cost) {
+                  foundPlansCurLevel.update(joinPlan.get.itemIds, joinPlan.get)
+                }
+              }
+            }
+          }
+        }
+
+        k += 1
+      }
+    }
+  }
+
+  /** Build a new join node. */
+  private def buildJoin(curJoinPlan: JoinPlan, otherJoinPlan: JoinPlan): Option[JoinPlan] = {
+    // Check if these two nodes are inner joinable. We consider cartesian product very
+    // costly, thus exclude such plans. This also helps us to reduce the search space.
+    val curPlan = curJoinPlan.plan
+    val otherPlan = otherJoinPlan.plan
+    val joinCond = conditions
+      .filterNot(l => canEvaluate(l, curPlan))
+      .filterNot(r => canEvaluate(r, otherPlan))
+      .filter(e => e.references.subsetOf(curPlan.outputSet ++ otherPlan.outputSet))
+
+    if (joinCond.nonEmpty) {
+      val curPlanStats = curPlan.stats(conf)
+      val otherPlanStats = otherPlan.stats(conf)
+      if (curPlanStats.rowCount.nonEmpty && otherPlanStats.rowCount.nonEmpty) {
+        // Now both curPlan and otherPlan become intermediate joins, so the cost of the
+        // new join should also include their costs.
+        val cost = curJoinPlan.cost + otherJoinPlan.cost +
```

Comment:
- Join estimation give us output of a LogicalPlan `p`(its size and cardinality), denoted as `T(p)` for simplicity.
- Cost of the plan is: `Cost(p) = sum(T(join))`, that is, the sum of output of all **intermediate** joins inside the tree of `p`, thus not include leaves or the root of tree. E.g. if p is (A J B) J C, A/B/C are leaves, then cost(p) is just T(A J B), rather than T(p) + T(A J B).

Then, in the above example, suppose A/B/C are all joins, A J B is just another join, so we denote it as D.
```
Costs(CURRENT) = Costs(D) = sum(T(join1)), join1 doesn't contain D
Costs(OTHER) = Costs(C) = sum(T(join2)), join2 doesn't contain C
Costs(NEW_JOIN) = Costs(CURRENT) + Costs(OTHER) + T(D) + T(C)
```
So there's no duplication here.
Maybe it would be less confusing if the code looks like:
```
val cost = Cost(
  curJoinPlan.cost.rows + curPlanStats.rowCount.get + otherJoinPlan.cost.rows + otherPlanStats.rowCount.get, 
  curJoinPlan.cost.size + curPlanStats.size + otherJoinPlan.cost.size + otherPlanStats.size)
```